File: src/leetcode/0017-Letter-Combinations-of-a-Phone-Number/solution2.py
'''
@Author: shaqsnake
@Email: shaqsnake@gmail.com
@Date: 2019-05-09 13:47:55
@LastEditTime: 2019-08-14 10:54:43
@Description: 
'''
#coding=utf-8
"""
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.



Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
"""
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        >>> Solution().letterCombinations('')
        []
        >>> Solution().letterCombinations('2')
        ['a', 'b', 'c']
        >>> Solution().letterCombinations('23')
        ['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']
        """
        if len(digits) == 0:
            return []
        d = {'1': '',
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz',
            '0': ' '
            }
        res = list(d[digits[0]])
        for i in range(1, len(digits)):
            res = self.prod_str(res, d[digits[i]])
        return res

    def prod_str(self, a, b):
        res = []
        for i in a:
            for j in b:
                res.append(i+j)
        return res

    # 利用python的reduce方法
    def letterCombinations_reduce(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        >>> Solution().letterCombinations_reduce('')
        []
        >>> Solution().letterCombinations_reduce('2')
        ['a', 'b', 'c']
        >>> Solution().letterCombinations_reduce('23')
        ['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']
        """
        if len(digits) == 0:
            return []
        d = {'1': '', '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz', '0': ' '}
        return reduce(lambda acc, digit: [x+y for x in acc for y in d[digit]], digits, [''])

    # 最简洁的写法
    def letterCombinations_oneline(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        >>> Solution().letterCombinations_online('')
        []
        >>> Solution().letterCombinations_online('2')
        ['a', 'b', 'c']
        >>> Solution().letterCombinations_online('23')
        ['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']
        """
        d = {'1': '', '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl', '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz', '0': ' '}
        return [x+y for x in d.get(digits[:1], '')
                    for y in self.letterCombinations_online(digits[1:]) or ['']] or []

if __name__ == '__main__':
    import doctest
    doctest.testmod(verbose=True)

File: src/leetcode/0547-Friend-Circles/solution2.cpp
// Copyright (c) 2019 shaqsnake. All rights reserved.
/**
 * @Author: shaqsnake
 * @Email: shaqsnake@gmail.com
 * @Date: 2019-06-11 18:23:09
 * @LastEditTime: 2019-08-26 16:41:59
 * @Description: 547. Friend Circles
 */
class Solution {
public:
    int findCircleNum(vector<vector<int>> &M) {
        if (M.empty() || M[0].empty())
            return 0;

        int n = M.size();
        for (int i = 1; i <= n; i++)
            p[i] = i;

        for (int i = 0; i < n; i++)
            for (int j = 0; j < i; j++)
                if (M[i][j] == 1)
                    uni(i + 1, j + 1);

        int res = 0;
        for (int i = 1; i <= n; i++)
            if (i == p[i])
                res++;
        return res;
    }

private:
    int p[1210];

    void uni(int x, int y) {
        int px = find(x), py = find(y);
        if (px != py)
            p[px] = py;
    }

    int find(int x) {
        if (x != p[x])
            p[x] = find(p[x]);
        return p[x];
    }
};